src/ore/src/id_gen.rs

```diff
@@ -97,8 +108,155 @@ where
+    /// used.
+    pub const fn new_static(id: T) -> Self {
+        IdHandle::Static(id)
+    }
+
+    /// Returns the ID associated with this [`IdHandle`].
+    pub fn val(&self) -> T {
+        match self {
+            IdHandle::Static(id) => *id,
+            IdHandle::Dynamic(inner) => *inner.id(),
+        }
+    }
+
+    /// Allocates a new ID and returns an owned [`IdHandle`] that can be cloned.
+    fn new(allocator: &IdAllocator<T>) -> Option<Self> {
+        let inner = Arc::new(internal::IdHandleInner::new(allocator)?);
+        Some(IdHandle::Dynamic(inner))
+    }
+}
+
+impl<T> Borrow<T> for IdHandle<T>
+where
+    T: From<u8> + AddAssign + PartialOrd + Copy,
+{
+    fn borrow(&self) -> &T {
+        match self {
+            IdHandle::Static(id) => id,
+            IdHandle::Dynamic(inner) => inner.id(),
+        }
+    }
+}
+
+impl<T> fmt::Display for IdHandle<T>
+where
+    T: fmt::Display + From<u8> + AddAssign + PartialOrd + Copy,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.val().fmt(f)
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<T> serde::Serialize for IdHandle<T>
+where
+    T: serde::Serialize + From<u8> + AddAssign + PartialOrd + Copy,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.val().serialize(serializer)
+    }
+}
+
+mod internal {
+    use std::ops::AddAssign;
+
+    use crate::id_gen::IdAllocator;
+
+    #[derive(Debug)]
+    pub struct IdHandleInner<T>
+    where
+        T: From<u8> + AddAssign + PartialOrd + Copy,
+    {
+        /// A handle to the [`IdAllocator`] used to allocated the provided id.
+        allocator: IdAllocator<T>,
+        /// The actual ID that was allocated.
+        id: T,
+    }
+
+    impl<T> IdHandleInner<T>
+    where
+        T: From<u8> + AddAssign + PartialOrd + Copy,
+    {
+        pub fn new(allocator: &IdAllocator<T>) -> Option<Self> {
+            let id = allocator.alloc()?;
+            Some(IdHandleInner {
+                allocator: allocator.clone(),
+                id,
+            })
+        }
+
+        pub fn id(&self) -> &T {
+            &self.id
+        }
+    }
+
+    impl<T> PartialEq for IdHandleInner<T>
+    where
+        T: PartialEq + From<u8> + AddAssign + PartialOrd + Copy,
+    {
+        fn eq(&self, other: &Self) -> bool {
+            self.id == other.id
+        }
+    }
+    impl<T> Eq for IdHandleInner<T> where T: PartialEq + From<u8> + AddAssign + PartialOrd + Copy {}
+
+    impl<T> PartialOrd for IdHandleInner<T>
+    where
+        T: PartialOrd + From<u8> + AddAssign + PartialOrd + Copy,
+    {
+        fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+            self.id.partial_cmp(&other.id)
+        }
+    }
+
+    impl<T> Ord for IdHandleInner<T>
+    where
+        T: Ord + From<u8> + AddAssign + PartialOrd + Copy,
+    {
+        fn cmp(&self, other: &Self) -> std::cmp::Ordering {
```

Comment:
For my learning: my read of the rust docs about the PartialOrd trait is that it doesn't implement the `cmp` fn. Where is it coming from? Also why do we need this?

Reply:
This is the `Ord` trait which has `fn cmp` as a required method. `PartialOrd` is implemented directly above. We need this because `BTreeMap` requires that keys implement `Ord`, and `Ord` requires we implement `Eq` and `PartialOrd`. We manually implement these as opposed to deriving them because `IdAllocator` doesn't implement them, nor would we want the `IdAllocator` to be considered when comparing `IdHandle`s, it should be based on the id itself.